Compile a workspace

Version: Deno 2.1.3

Hi,

Given the following code:
```ts
// lib-a/main.ts
export const a = "Hello from lib-a";
```
```json
// lib-a/package.json
{
  "name": "@my-scope/lib-a",
  "type": "module",
  "version": "1.0.0",
  "main": "main.ts",
  "exports": "./main.ts"
}
```
```ts
// lib-b/main.ts
import { a } from "@my-scope/lib-a";

console.log(a);
```
```json
// lib-b/package.json
{
  "name": "@my-scope/lib-b",
  "type": "module",
  "version": "1.0.0",
  "main": "main.ts",
  "exports": "./main.ts"
}
```
```
// deno.json
{
  "workspace": ["./lib-a", "./lib-b"]
}
```

A small repo with the code above: https://github.com/tjoskar/deno-workspace-bug-maybe

Running `deno run lib-b/main.ts` works fine but if I run `deno check lib-b/main.ts` I get the error:
```
error: Failed resolving types. [ERR_TYPES_NOT_FOUND] Could not find types for 'file:///deno-compile-workspace/lib-a/index.js' imported from 'file:///deno-compile-workspace/lib-b/main.ts'
    at file:///deno-compile-workspace/lib-b/main.ts:1:19
```

And if I run `deno compile --no-check --output test lib-b/main.ts` and then `./test` I get the same error (ish):
```
error: [ERR_MODULE_NOT_FOUND] Cannot find module 'file:///var/folders/n3/3d2w2d152hb6fs7d1hkwyvjc0000gn/T/deno-compile-test/deno-compile-workspace/lib-a/index.js' imported from 'file:///var/folders/n3/3d2w2d152hb6fs7d1hkwyvjc0000gn/T/deno-compile-test/deno-compile-workspace/lib-b/main.ts'
```

Am I missing something? Or isn't this supported?

It works fine if I replace both `package.json` with corresponding deno.json files, but I can't do that for several reasons right now. 

Thanks! And only love to Deno!

--- a/cli/standalone/binary.rs
+++ b/cli/standalone/binary.rs
@@ -394,6 +394,13 @@ impl<'a> DenoCompileBinaryWriter<'a> {
     for path in exclude_paths {
       vfs.add_exclude_path(path);
     }
+    // Embed the workspace package.json files so the standalone binary's node
+    // resolver can read their "exports" (and other) fields at runtime. Without
+    // this, resolving a workspace member by its package name falls back to
+    // legacy `index.js` resolution instead of honoring the package's exports.
+    for pkg_json in self.cli_options.workspace().package_jsons() {
+      vfs.add_path(&pkg_json.path)?;
+    }
     let progress_bar = ProgressBar::new(ProgressBarStyle::ProgressBars);
     let npm_snapshot = match &self.npm_resolver {
       CliNpmResolver::Managed(managed) => {
@@ -983,9 +990,6 @@ impl<'a> DenoCompileBinaryWriter<'a> {
       }
       CliNpmResolver::Byonm(_) => {
         maybe_warn_different_system(&self.npm_system_info);
-        for pkg_json in self.cli_options.workspace().package_jsons() {
-          builder.add_path(&pkg_json.path)?;
-        }
         let _progress =
           progress_bar.update_with_prompt(ProgressMessagePrompt::Compile, "");
         // traverse and add all the node_modules directories in the workspace

--- a/tests/specs/compile/workspace_package_json_exports/__test__.jsonc
+++ b/tests/specs/compile/workspace_package_json_exports/__test__.jsonc
@@ -1,0 +1,22 @@
+{
+  "tempDir": true,
+  "steps": [{
+    "if": "unix",
+    "args": "compile --no-check --output main main.ts",
+    "output": "[WILDCARD]"
+  }, {
+    "if": "unix",
+    "commandName": "./main",
+    "args": [],
+    "output": "main.out"
+  }, {
+    "if": "windows",
+    "args": "compile --no-check --output main.exe main.ts",
+    "output": "[WILDCARD]"
+  }, {
+    "if": "windows",
+    "commandName": "./main.exe",
+    "args": [],
+    "output": "main.out"
+  }]
+}

--- a/tests/specs/compile/workspace_package_json_exports/deno.json
+++ b/tests/specs/compile/workspace_package_json_exports/deno.json
@@ -1,0 +1,3 @@
+{
+  "workspace": ["./libs/lib"]
+}

--- a/tests/specs/compile/workspace_package_json_exports/libs/lib/package.json
+++ b/tests/specs/compile/workspace_package_json_exports/libs/lib/package.json
@@ -1,0 +1,8 @@
+{
+  "name": "@repo/lib",
+  "version": "1.0.0",
+  "exports": {
+    ".": "./dist/index.js",
+    "./util": "./dist/util.js"
+  }
+}

--- a/tests/specs/compile/workspace_package_json_exports/main.out
+++ b/tests/specs/compile/workspace_package_json_exports/main.out
@@ -1,0 +1,2 @@
+hello from @repo/lib
+3

--- a/tests/specs/compile/workspace_package_json_exports/main.ts
+++ b/tests/specs/compile/workspace_package_json_exports/main.ts
@@ -1,0 +1,8 @@
+// Importing a non-Deno workspace member (package.json with an "exports" map) by
+// its package name must honor the exports map in the compiled binary, both for
+// the root export and a subpath export. Regression test for #28926.
+import { greet } from "@repo/lib";
+import { sum } from "@repo/lib/util";
+
+console.log(greet());
+console.log(sum(1, 2));

--- a/tests/specs/compile/workspace_package_json_exports_ts/__test__.jsonc
+++ b/tests/specs/compile/workspace_package_json_exports_ts/__test__.jsonc
@@ -1,0 +1,22 @@
+{
+  "tempDir": true,
+  "steps": [{
+    "if": "unix",
+    "args": "compile --no-check --output main lib-b/main.ts",
+    "output": "[WILDCARD]"
+  }, {
+    "if": "unix",
+    "commandName": "./main",
+    "args": [],
+    "output": "main.out"
+  }, {
+    "if": "windows",
+    "args": "compile --no-check --output main.exe lib-b/main.ts",
+    "output": "[WILDCARD]"
+  }, {
+    "if": "windows",
+    "commandName": "./main.exe",
+    "args": [],
+    "output": "main.out"
+  }]
+}

--- a/tests/specs/compile/workspace_package_json_exports_ts/deno.json
+++ b/tests/specs/compile/workspace_package_json_exports_ts/deno.json
@@ -1,0 +1,3 @@
+{
+  "workspace": ["./lib-a", "./lib-b"]
+}

--- a/tests/specs/compile/workspace_package_json_exports_ts/lib-a/main.ts
+++ b/tests/specs/compile/workspace_package_json_exports_ts/lib-a/main.ts
@@ -1,0 +1,1 @@
+export const a: string = "Hello from lib-a";

--- a/tests/specs/compile/workspace_package_json_exports_ts/lib-a/package.json
+++ b/tests/specs/compile/workspace_package_json_exports_ts/lib-a/package.json
@@ -1,0 +1,7 @@
+{
+  "name": "@my-scope/lib-a",
+  "type": "module",
+  "version": "1.0.0",
+  "main": "main.ts",
+  "exports": "./main.ts"
+}

--- a/tests/specs/compile/workspace_package_json_exports_ts/lib-b/main.ts
+++ b/tests/specs/compile/workspace_package_json_exports_ts/lib-b/main.ts
@@ -1,0 +1,7 @@
+// A workspace member (package.json) importing another workspace member by its
+// package name, where the "exports" map points at a TypeScript file. The
+// compiled binary must honor the exports map and transpile the target.
+// Regression test for #27315.
+import { a } from "@my-scope/lib-a";
+
+console.log(a);

--- a/tests/specs/compile/workspace_package_json_exports_ts/lib-b/package.json
+++ b/tests/specs/compile/workspace_package_json_exports_ts/lib-b/package.json
@@ -1,0 +1,7 @@
+{
+  "name": "@my-scope/lib-b",
+  "type": "module",
+  "version": "1.0.0",
+  "main": "main.ts",
+  "exports": "./main.ts"
+}

--- a/tests/specs/compile/workspace_package_json_exports_ts/main.out
+++ b/tests/specs/compile/workspace_package_json_exports_ts/main.out
@@ -1,0 +1,1 @@
+Hello from lib-a

